### PR TITLE
WIP fix: bugs in generic contracts on corner cases (bug 1)

### PR DESCRIPTION
**Pull request by ZhectorSM**
Fixed bug 1: method checkShadow in the class ContractVerifier
Related to #2732

### CODE REVIEW

**Comment by ZhectorSM** [comment]
There is an error with the spaces, I tried to fix it but it didn't work.
I don't know why.

File: src/main/java/spoon/ContractVerifier.java [hunk 1]
```diff
@@ -202,7 +202,8 @@ public void checkShadow() {
 			@Override
 			public void scan(CtElement element) {
 				if (element != null && CtShadowable.class.isAssignableFrom(element.getClass())) {
-					assertFalse(((CtShadowable) element).isShadow());
+                                        assertFalse(((CtShadowable) element).isShadow());
```
Comment (nharrand) [inline src/main/java/spoon/ContractVerifier.java]:
```suggestion
					assertFalse(((CtShadowable) element).isShadow());
					return;
```

File: src/main/java/spoon/ContractVerifier.java [hunk 2]
```diff
@@ -202,7 +202,8 @@ public void checkShadow() {
 			@Override
 			public void scan(CtElement element) {
 				if (element != null && CtShadowable.class.isAssignableFrom(element.getClass())) {
-					assertFalse(((CtShadowable) element).isShadow());
+                                        assertFalse(((CtShadowable) element).isShadow());                                        
+                                    return;
```
Comment (nharrand) [inline src/main/java/spoon/ContractVerifier.java]:
```suggestion
```

**Comment by nharrand** [comment]
It seems to be just a small issues with tabulation versus spaces (and line should not ennd with white characters).

**Comment by nharrand** [comment]
By the way, if you want to check if there are styling issues locally, you can run:

```sh
mvn checkstyle:checkstyle
```

**Comment by ZhectorSM** [comment]
Perfect!
Thank you for your help.

File: src/main/java/spoon/ContractVerifier.java [hunk 3]
```diff
@@ -203,6 +203,7 @@ public void checkShadow() {
 			public void scan(CtElement element) {
 				if (element != null && CtShadowable.class.isAssignableFrom(element.getClass())) {
 					assertFalse(((CtShadowable) element).isShadow());
+					return;
```
Comment (nharrand) [inline src/main/java/spoon/ContractVerifier.java]: If you do add this `return` here, you prevent the scanner (here `checkShadow`) to visitit descendent of a `CtElement` if it is a class that implement `CtShadowable`. As we run contract versifier on a root package (an instance of `CtPackage` which does implement `CtShadowable`), this patch would stop the visit immediatly, i.e. it stop verifying the contract. While it does make the test pass, it basically remove its body.

I think the root of the bug (at least in the case of `fieldaccesses/testclasses/Tacos`) is that `typeDeclaration` line 218 is null for a reference pointing to `Foo.SubInner`. 
(And AFAIU it should not be as we have the sources for `spoon.test.fieldaccesses.testclasses.internal.Bar.Inner.SubInner`, and `Foo` extends `Bar.Inner`, so  `Foo.SubInner` should exist and be correctly resolved.)

**Pull request closed without merging**